### DISCUSSION TITLE
Allow Ant theme customization

### DISCUSSION THREAD

**Discussion by Radiergummi** [lowdefy/lowdefy#678] [Ideas and feature requests]
As far as I see, there's no elegant way of changing the CSS to fit a corporate design, for example, other than adding an external style sheet, and missing out of all the goodness a modern web stack provides.  
Ideally, there'd be a global configuration section to configure the Ant Design theme as used by Lowdefy. Is that possible at all? Does this make sense? I'm fine with the default theme for the moment, but having more control over colors would help some of our employees, for example.

**Comment by Gervwyk** [comment]
Well, this is a sensitive topic 😉 I really really want to make this happen but to get it right is tricky according to my research (about 2 years ago so maybe a lot has changed). I'm not the most talented front-end guy, so please if there are any experts in the community they should jump in and help with advice 🙏

So antd is amazing in terms of customization at build time. They have an extensive list of theme variables that one would be able to configure and make app theme changes easy: 
https://github.com/ant-design/ant-design/blob/master/components/style/themes/default.less
https://ant.design/docs/react/customize-theme

The complication is that Lowdefy uses webpack module federation to split blocks into separate build bundles / packages which are loaded at runtime. Now because antd uses less to compile css at build time, modifying it at runtime is a little more complex: 
https://pro.ant.design/blog/change-theme
https://github.com/mzohaibqc/antd-theme-webpack-plugin

One idea was to replace the antd theme vars with css vars so that less compiles or replaces values with css variables. This breaks because less has a few functions which do calculations on things like colors etc. So probably not possible to bypass / overwrite those methods.

The second approach which might be more realistic, not to compile the css, but serve the less files to the client then run less on the browser. The downside would be a performance hit, which is not ideal. And also I've not seen anyone try this with antd.

I'm not really a fan of the examples of live theme editing which only changes colors. This method finds and replaces colors in the css config. which is then limited to what you can find and replace, so other vars would not be an option with this approach. There was another demo where someone gets it right and could edit all vars. I tried finding it now, but couldn't. I'll search again tomorrow.

A different option is to compile the css inside the lowdefy build process, and then serve css from the app file server. This is the approach I think we should investigate further. Since every app can then compile it's css and hos the css assets on the lowdefy server. 

What I did not yet investigate is how to only compile the css bundle for antd, and if this css bundle is 100% separate from the js assets, which I think it should be since it's less and not inline, at least for all vars. Another consideration is if we can make one build for all ant blocks, or maybe just for the blocks used on a page, or how the css files could be split.

**Reply by Radiergummi** [reply]
> A different option is to compile the css inside the lowdefy build process, and then serve css from the app file server. This is the approach I think we should investigate further. Since every app can then compile it's css and hos the css assets on the lowdefy server.

That sounds great. Compiling during the build, then statically serving a cached version of the application CSS is what the SSR frameworks like Next are doing too, I think.

> Another consideration is if we can make one build for all ant blocks, or maybe just for the blocks used on a page, or how the css files could be split.

Could Webpack module federation be leveraged to serve block CSS, too? Would be pretty cool if blocks could provide their own stylesheets (possibly as CSS modules, even?) that layer on top of the base Lowdefy styles.

**Reply by Gervwyk** [reply]
I think this could be done. It's going to take some experimentation. So ideally we'd like to specify the modified vars in the config:
```yaml
name: App
lowdefy: 3.18.1
app:
  css:
    antdVars:
      '@primary-color': red
      '@border-radius-base': 4px
      ...
  html:
    appendHead:
      ...
```
These settings then need to be loaded into a webpack config like so: 
https://github.com/lowdefy/lowdefy/blob/12e3a2e62254e6f385cbe73f92ee915a5a29e430/packages/layout/webpack.config.js#L70
Then run webpack build for the server and only build the css assets. What I'm not sure about is how to setup the files so that webpack knows what css for what blocks to include in the build. I'm pretty sure the css for all antd blocks will need to be built with the same vars. 

Any ideas on how this could work?
